librz/core/cmp.c

```diff
@@ -0,0 +1,465 @@
+// SPDX-FileCopyrightText: 2021 Dhruv Maroo <dhruvmaru007@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_cmp.h>
+#include <rz_asm.h>
+#include <rz_list.h>
+
+/**
+ * \brief Compare memory at \p addr1 with the memory at \p addr2
+ *
+ * \param core Current RzCore instance
+ * \param addr1 address to read data from
+ * \param addr2 address to read data from
+ * \param len Number of bytes to compare
+ * \return RzCompareData* A pointer to RzCompareData comparison
+ */
+RZ_API RZ_OWN RzCompareData *rz_core_cmp_mem_mem(RzCore *core, ut64 addr1, ut64 addr2, ut32 len) {
+	rz_return_val_if_fail(core, NULL);
+
+	ut8 *buf1 = malloc(len * sizeof(ut8));
+	ut8 *buf2 = malloc(len * sizeof(ut8));
+	if (!buf1 || !buf2) {
+		goto error_goto;
+	}
+	if (rz_io_nread_at(core->io, addr1, buf1, len) == -1 || rz_io_nread_at(core->io, addr2, buf2, len) == -1) {
+		RZ_LOG_ERROR("Cannot read at provided addresses: 0x%" PFMT64x " 0x%" PFMT64x "\n", addr1, addr2);
+		goto error_goto;
+	}
+	RzCompareData *cmp = RZ_NEW0(RzCompareData);
+	if (!cmp) {
+		goto error_goto;
+	}
+	cmp->len = len;
+	cmp->data1 = buf1;
+	cmp->addr1 = addr1;
+	cmp->data2 = buf2;
+	cmp->addr2 = addr2;
+	cmp->same = rz_mem_eq(cmp->data1, cmp->data2, len);
+	return cmp;
+
+error_goto:
+	free(buf1);
+	free(buf2);
+	return NULL;
+}
+
+/**
+ * \brief Compare mem at \p addr with data \p data
+ *
+ * \param core Current RzCore instance
+ * \param addr address to read data from
+ * \param data Data to be compared
+ * \param len Number of bytes to compare
+ * \return RzCompareData* A pointer to RzCompareData comparison (the data1 corresponds to the data at addr (and addr1 = addr) and data2 is the str data (and addr2 = UT32_MAX))
+ */
+RZ_API RZ_OWN RzCompareData *rz_core_cmp_mem_data(RzCore *core, ut64 addr, RZ_NONNULL const ut8 *data, ut32 len) {
+	rz_return_val_if_fail(core && data, NULL);
+
+	ut8 *buf1 = malloc(len * sizeof(ut8));
+	if (!buf1) {
+		RZ_LOG_ERROR("Cannot read at address: 0x%" PFMT64x "\n", addr);
+		goto error_goto;
+	}
+	if (rz_io_nread_at(core->io, addr, buf1, len) == -1) {
+		goto error_goto;
+	}
+	RzCompareData *cmp = RZ_NEW0(RzCompareData);
+	if (!cmp) {
+		goto error_goto;
+	}
+	cmp->len = len;
+	cmp->data1 = buf1;
+	cmp->addr1 = addr;
+	cmp->data2 = rz_mem_dup(data, len);
+	cmp->addr2 = UT64_MAX;
+	cmp->same = rz_mem_eq(cmp->data1, cmp->data2, len);
+	return cmp;
+
+error_goto:
+	free(buf1);
+	return NULL;
+}
+
+/**
+ * \brief Print a comparison \p cmp according to the print mode \p mode
+ *
+ * \param core Current RzCore instance
+ * \param cmp RzCompareData instance to be printed
+ * \param mode Mode to be used (options: default, diff, json)
+ * \return int Number of lines/diffs printed (-1 if failed)
+ */
+RZ_API int rz_core_cmp_print(RzCore *core, RZ_NONNULL const RzCompareData *cmp, RzOutputMode mode) {
```

Comment:
Maybe you already answered somewhere, but did you consider using `RzCmdStateOutput` (and RZ_CMD_DESC_TYPE_ARGV_STATE in the command instead of ARGV_MODES) ?

Reply:
Sure, I'll do that. The only reason why I didn't do it before was because this function was creating its own JSON object instead of using the one which comes with `RZ_CMD_DESC_TYPE_ARGV_STATE`, so using `ARGV_STATE` was useless. However, I think I should refactor the definition to use the JSON object provided by the `ARGV_STATE`.

Reply:
> because to me it feels easier to just pass around an enum instead of creating a RzCmdStateOutput object and then pass it

You don't need to create a `RzCmdStateOutput` in most cases because it comes directly from the command handler if you use `RZ_CMD_DESC_TYPE_ARGV_STATE`. It actually isn't strictly related to "command handling" but more to "output handling" as it makes sure that you have all the right data structures correctly initialized for the right mode you want to output (e.g. if output mode is RZ_OUTPUT_MODE_TABLE you'd have the RzTable in ->d.t, etc.).

BTW, there are functions like `rz_cmd_state_output_init` and others to quickly create a `RzCmdStateOutput` object, though in your cases you probably don't even need it.

> Also, it is possible that in future this might be called by other functions completely unrelated to command handling

I doubt that's ever going to be the case considering that the `_print` function directly prints to RzCons, thus something like Cutter can hardly calls this function to get any useful results (apart from just printing stuff).

An added bonus is that e.g. `cmd_cmp_bytes`, `cmd_cmp_addr` and other functions where you are now using `rz_core_cmp_print(...., RZ_CMD_STATE_OUTPUT_STANDARD)` could become `rz_core_cmp_print(...., state)` and you would get for free JSON support also for those commands (`ca`, `cb` and others) just by adding the modes to the cmd_cmp.yaml file.